### DISCUSSION TITLE
Vuser connections (PostgreSQL)

### DISCUSSION THREAD

**Discussion by miashaee** [TPC-Council/HammerDB#393] [Q&A] [answered]
Hi, I have a general question about how connections are formed to PostgreSQL for vusers for TPCC runs in a PostgreSQL, I am running 1000000 transactions with up to 60 vusers and I only see one connection on my PostgreSQL database at any time for HammerDB no matter the number of virtual users. How exactly does HammerDB connect to PostgreSQL with vusers? Thanks.

**Comment by sm-shaw** [comment]
you don't mention the OS you are using.
Connections are made with libpq the native C library.
This blog post https://www.hammerdb.com/blog/uncategorized/why-tcl-is-700-faster-than-python-for-database-benchmarking/ is a good illustration of how you can see a difference in the number of virtual users and the workload.
Note HammerDB is multithreaded so the virtual users themselves are independent threads under one process.

**Reply by miashaee** [reply]
I am utilizing Centos 7.9 on PostgreSQL, wondering how that impacts how connections are formed as I only see one connection for the up to 60 vusers that are being utilized. Does each vuser act as a separate transaction on the same connection? Like one user transaction after another (serially as opposed to in parallel)?

**Comment by sm-shaw** [comment, marked as answer]
There does seem to be some confusion here, it is well worth re-reading the blog post and searching for the word "parallel", that is the crucial concept that HammerDB creates multiple Virtual Users that connect to the database and run the workload in parallel. (Many tools aren't able to do this, HammerDB does).

So as an example, here 10 Virtual Users are created, we also create a monitor virtual user so that is 11 in total.

![10vuser](https://user-images.githubusercontent.com/38044085/179244099-765a8862-9277-4004-b068-d32fa90d7125.png)

 For the command line vuset vu 10 and vucreate e.g.

```
hammerdb>vuset vu 10

hammerdb>vucreate
Vuser 1 created MONITOR - WAIT IDLE
Vuser 2 created - WAIT IDLE
Vuser 3 created - WAIT IDLE
Vuser 4 created - WAIT IDLE
Vuser 5 created - WAIT IDLE
Vuser 6 created - WAIT IDLE
Vuser 7 created - WAIT IDLE
Vuser 8 created - WAIT IDLE
Vuser 9 created - WAIT IDLE
Vuser 10 created - WAIT IDLE
Vuser 11 created - WAIT IDLE
11 Virtual Users Created with Monitor VU

hammerdb>

```
Once we start running (vurun in the CLI) if we check in the database we can see that 11 sessions are created:

```
postgres=# select count(*) from pg_stat_activity where usename = 'tpcc';
 count 
-------
    11
(1 row)
```
and as postgres is a process based database we can see those 11 sessions as processes:

```
postgres       2951    2637  0 15:20 ?        00:00:00 postgres: tpcc tpcc 127.0.0.1(59578) idle
postgres       2952    2637 30 15:20 ?        00:00:42 postgres: tpcc tpcc 127.0.0.1(59580) SELECT
postgres       2953    2637 24 15:20 ?        00:00:34 postgres: tpcc tpcc 127.0.0.1(59582) idle
postgres       2954    2637 31 15:20 ?        00:00:43 postgres: tpcc tpcc 127.0.0.1(59584) SELECT
postgres       2955    2637 31 15:20 ?        00:00:42 postgres: tpcc tpcc 127.0.0.1(59586) BIND
postgres       2956    2637 29 15:20 ?        00:00:40 postgres: tpcc tpcc 127.0.0.1(59588) SELECT
postgres       2957    2637 30 15:20 ?        00:00:41 postgres: tpcc tpcc 127.0.0.1(59590) idle
postgres       2958    2637 29 15:20 ?        00:00:40 postgres: tpcc tpcc 127.0.0.1(59592) SELECT
postgres       2959    2637 29 15:20 ?        00:00:40 postgres: tpcc tpcc 127.0.0.1(59594) idle
postgres       2960    2637 30 15:20 ?        00:00:40 postgres: tpcc tpcc 127.0.0.1(59596) idle
postgres       2961    2637 29 15:20 ?        00:00:40 postgres: tpcc tpcc 127.0.0.1(59598) idle
```

Note you won't see 10 processes on the HammerDB client side as those 10 Virtual Users are running as threads inside the HammerDB process. 
For advanced users you can even create more than one session per virtual user called event-driven scaling, however by default the Virtual Users you create, creates a session and they all run in _parallel_ as designed. 

From this point it is definitely worth reading the documentation and the blog to understand some of the more basic concepts first.
